thorlcr/thorutil/thbuf.cpp

```diff
@@ -606,6 +608,549 @@ class CSmartRowInMemoryBuffer: public CSimpleInterface, implements ISmartRowBuff
+                        {
+                            trace("READ: setting currentTempFileEndRow: %" RCPF "u", currentTempFileEndRow.load());
+                            break;
+                        }
+                        else if (outputFileEndRowMarkers.empty())
+                        {
+                            trace("READ: setting currentTempFileEndRow: unbounded (2)");
+                            currentTempFileEndRow = (rowcount_t)-1; // unbounded for now, writer will set when it knows
+                            break;
+                        }
+                    }
+                }
+                else
+                {
+                    currentTempFileEndRow = (rowcount_t)-1; // unbounded for now, writer will set when it knows
+                    trace("READ: setting currentTempFileEndRow: unbounded");
+                }
+            }
+        }
+        if (grouped)
+        {
+            bool eog;
+            inputStream->read(sizeof(bool), &eog);
+            if (eog)
+                return nullptr;
+        }
+        RtlDynamicRowBuilder rowBuilder(allocator);
+        size32_t sz = deserializer->deserialize(rowBuilder, inputDeserializerSource);
+        const void *row = rowBuilder.finalizeRowClear(sz);
+        checkCurrentRow("S: ", row, nextInputRow);
+        return row;
+    }
+    void writeRowToStream(const void *row, size32_t rowSz)
+    {
+        if (!spilt)
+        {
+            spilt = true;
+            ActPrintLog(&activity, "Spilling to temp storage [file = %s]", baseTmpFilename.get());
+            createNextOutputStream();
+        }
+        if (grouped)
+        {
+            bool eog = (nullptr == row);
+            outputStream->put(sizeof(bool), &eog);
+            pendingFlushToDiskSz++;
+            if (nullptr == row)
+                return;
+        }
+        serializer->serialize(*outputStreamSerializer.get(), (const byte *)row);
+        pendingFlushToDiskSz += rowSz;
+    }
+    void checkReleaseQBlockReader()
+    {
+        if (readerWaitingForQ)
+        {
+            readerWaitingForQ = false;
+            moreRows.signal();
+        }
+    }
+    void checkReleaseReaderCommitBlocked()
+    {
+        if (readerWaitingForCommit)
+        {
+            readerWaitingForCommit = false;
+            moreRows.signal();
+        }
+    }
+    void handleInputComplete()
+    {
+        readState = rs_stopped;
+        if (flushWaiting)
+        {
+            flushWaiting = false;
+            flushWaitSem.signal();
+        }
+    }
+    bool checkFlushToDisk(size32_t threshold)
+    {
+        if (pendingFlushToDiskSz <= threshold)
+            return false;
+        rowcount_t _nextOutputRow = nextOutputRow.load();
+        trace("WRITE: Flushed to disk. nextOutputRow = %" RCPF "u", _nextOutputRow);
+        outputStream->flush();
+        currentTempFileSize += pendingFlushToDiskSz;
+        pendingFlushToDiskSz = 0;
```

Comment:
Call to update the current temp file usage?
activeOutput->noteSize(currentTempFileSize);